sycl/include/sycl/detail/kernel_launch_helper.hpp

```diff
@@ -251,20 +254,325 @@ struct KernelWrapper<
+        bool supported = p->deviceSupportForwardProgress(
+            prop.guarantee,
+            sycl::ext::oneapi::experimental::execution_scope::work_item,
+            prop.coordinationScope);
+        verifyDeviceHasProgressGuarantee(
+            supported, prop.guarantee,
+            sycl::ext::oneapi::experimental::execution_scope::work_item,
+            retval);
+      }
+    }
+
+    // Process work group scratch memory property.
+    {
+      if constexpr (PropertiesT::template has_property<
+                        sycl::ext::oneapi::experimental::
+                            work_group_scratch_size>()) {
+        throwIfGraphAssociated<syclex::detail::UnsupportedGraphFeatures::
+                                   sycl_ext_oneapi_work_group_scratch_memory>(
+            HasGraph);
+        auto WorkGroupMemSize = Props.template get_property<
+            sycl::ext::oneapi::experimental::work_group_scratch_size>();
+        retval.MKernelWorkGroupMemorySize = WorkGroupMemSize.size;
+      }
+    }
+
+    // Parse cluster properties.
+    {
+      constexpr std::size_t ClusterDim =
+          syclex::detail::getClusterDim<PropertiesT>();
+      if constexpr (ClusterDim > 0) {
+        throwIfGraphAssociated<
+            syclex::detail::UnsupportedGraphFeatures::
+                sycl_ext_oneapi_experimental_cuda_cluster_launch>(HasGraph);
+        auto ClusterSize = Props
+                               .template get_property<
+                                   syclex::cuda::cluster_size_key<ClusterDim>>()
+                               .get_cluster_size();
+        retval.MKernelUsesClusterLaunch = true;
+        retval.MKernelClusterDims = ClusterDim;
+        if (ClusterDim == 1) {
+          retval.MKernelClusterSize[0] = ClusterSize[0];
+        } else if (ClusterDim == 2) {
+          retval.MKernelClusterSize[0] = ClusterSize[0];
+          retval.MKernelClusterSize[1] = ClusterSize[1];
+        } else if (ClusterDim == 3) {
+          retval.MKernelClusterSize[0] = ClusterSize[0];
+          retval.MKernelClusterSize[1] = ClusterSize[1];
+          retval.MKernelClusterSize[2] = ClusterSize[2];
+        } else {
+          assert(ClusterDim <= 3 &&
+                 "Only 1D, 2D, and 3D cluster launch is supported.");
+        }
+      }
+    }
+
+    return retval;
+  }
+
+  /// Process kernel properties.
+  ///
+  /// Stores information about kernel properties into the handler.
+  ///
+  /// Note: it is important that this function *does not* depend on kernel
+  /// name or kernel type, because then it will be instantiated for every
+  /// kernel, even though body of those instantiated functions could be almost
+  /// the same, thus unnecessary increasing compilation time.
+  template <
+      bool IsESIMDKernel, typename PropertyProcessorPtrT,
+      typename PropertiesT = ext::oneapi::experimental::empty_properties_t>
+  static KernelLaunchPropertiesT processProperties(PropertiesT Props,
```

Comment:
I was thinking about inlining launch properties processing before, but it makes even less sense to have them separate now.